modules/ocf/manifests/ssl/lets_encrypt/dns.pp

```diff
@@ -10,24 +10,30 @@
     content => "${join($domains, ' ')} > ${title}",
   }
 
-  # Only run the dehydrated command to renew the cert if it is old enough (30
-  # days = 2592000 seconds left until it expires). dehydrated does this check
-  # itself, but this should help with puppet error spam a bit (for instance
-  # if Let's Encrypt is down for any reason) and mean that services can
-  # subscribe to this class to reload after new certs are obtained since this
-  # won't refresh unless dehydrated is re-run for some reason.
-  exec { "check ${title} cert expiration":
-    command => '/bin/true',
-    unless  => "openssl x509 -checkend 2592000 -noout -in /var/lib/lets-encrypt/certs/${title}/cert.pem",
-    user    => ocfletsencrypt,
-  } ~>
+  # Only run the dehydrated command to renew the cert if it is old enough, or
+  # missing some domains. dehydrated does the former check itself, but this
+  # should help with puppet error spam a bit (for instance if Let's Encrypt is
+  # down for any reason) and mean that services can subscribe to this class to
+  # reload after new certs are obtained since this won't refresh unless
+  # dehydrated is re-run for some reason.
+
+  $parsed_cert_info = parsejson($::le_cert_info)
+
+  $have_cert_info = $title in $parsed_cert_info
+  if $have_cert_info {
+    $cert_expires_soon = ($parsed_cert_info[$title]['days_to_expiration'] + 0) < 30
+    # we subtract out any domains that are in the cert, and see if any are left
+    $cert_has_all_domains = ($domains - $parsed_cert_info[$title]['cert_names']) =~ Array[String, 0, 0]
+  }
+
   exec { "obtain ${title} cert":
     # This exec can be notified to get it to run dehydrated again, even if the
     # cert will not expire soon.
     command     => '/usr/bin/dehydrated --cron --privkey /etc/ssl/lets-encrypt/le-account.key',
     user        => $owner,
-    refreshonly => true,
     require     => Package['dehydrated-hook-ddns-tsig'],
+    # if the cert's chillin, only exec on notification
+    refreshonly => $have_cert_info and !$cert_expires_soon and $cert_has_all_domains,
```

Comment:
Hmm, not sure if this is the logic we want here. If the first run fails to obtain a cert, I think these resources will be updated while the cert stays out of date. Then the second run, `refreshonly` will be set to true, but that says to only run if dependent resources are updated, which they will not be. 

Someone double check me on this though

Reply:
shouldn't it pull this info fresh from the cert on disk every time it runs? if dehydrated fails to acquire the new cert, the next time puppet runs, the fact will return info from the unchanged cert.

Reply:
Refreshonly is just a boolean that says whether or not we should only run the exec when the resource is notified. If the cert is out of date, or doesn't have all the domains, we should run the exec every time (refreshonly = false). Otherwise we should wait for notification (refreshonly = true).

Reply:
I see... if we get a new config or something, and then fail to obtain a cert, we still won't retry. Not sure how to handle that case.

Reply:
I think that's fine, it should be a rare case where we change the config, and definitely a much rarer case than the current system fails on.

The other way you could do this is have some kind of file or other state that indicates a previously failed run and always re-run if the prior run has failed, something like that. It's definitely not ideal to have to do that though.

Reply:
Yeah, this is an edge case. Maybe just leave a comment and dwai